### PR TITLE
[Kernel] Make FlashInfer CuTeDSL the default unquantized BF16 GEMM and drop skinny-GEMM dispatch

### PR DESCRIPTION
**Pull request by zyongye**
## Purpose

Follow-up to #50572, which integrated the FlashInfer BF16 CuTeDSL low-latency GEMM behind `--linear-backend=flashinfer_cutedsl`. This PR makes that kernel the default for unquantized BF16 linear layers and retires the hand-tuned skinny-GEMM shape dispatch it supersedes.

**AI assistance was used for this PR** (Claude Code). Every changed line was reviewed by me, and the tests, benchmarks and evals below were run on GB300.

### 1. FlashInfer CuTeDSL BF16 becomes the default

`dispatch_unquantized_gemm("auto")` now selects the CuTeDSL `mm_bf16` kernel wherever it is supported (sm100a, BF16, M ≤ 32), instead of only under an explicit `--linear-backend=flashinfer_cutedsl`. Quantized kernel selection is untouched — `_LINEAR_BACKEND_KERNEL_MAP` and the `KernelConfig.linear_backend` default (`"auto"`) are unchanged.

### 2. Cleanups

- Removed `has_flashinfer_bf16_gemm` and `is_flashinfer_bf16_gemm_supported`. `requirements/cuda.txt` pins `flashinfer-python==0.6.18`, which always exposes `mm_bf16` and `mm_bf16.is_backend_supported`, so the `getattr`/`callable` probing was dead. `is_flashinfer_cutedsl_bf16_gemm_supported()` now checks the things that can actually vary: CUDA, FlashInfer present, CuTe DSL importable, sm100a.
- Collapsed the single-entry `_FLASHINFER_BF16_BACKENDS` registry (dataclass + lookup + `ValueError` path) and dropped the now-constant `vllm_backend` argument from the `cuda_flashinfer_bf16_gemm` custom op.
- `dispatch_unquantized_gemm` is now cached, and CPU activations short-circuit to `torch.nn.functional.linear` rather than failing to dispatch a CUDA-only op.

### 3. Skinny GEMM is no longer auto-dispatched

The per-shape skinny-GEMM dispatch is removed from every model path:

- Deleted `qwen4_exp/nvidia/low_latency_gemm.py` (skinny-only) and its install calls. The ROCm half of Qwen4Exp is deliberately untouched — its `low_latency_gemm.py` is already a no-op stub, so ROCm behaviour is unchanged and no file under `amd/` is modified.
- Kimi-K3 and GLM-5.2 keep their `dsv3_fused_a_gemm` shape dispatch — see the benchmark below — but lose the `"cute"` backend, the `cute_configs`/`residual_configs` tables, `select_kimi_k3_backend`, `try_low_latency_gemm`, `apply_with_residual` (dead: nothing in `vllm/` called it) and `KimiK3LowLatencyEmbeddingMethod` (unreachable once the LM-head shapes were skinny-only).
- Removed the explicit `try_low_latency_gemm` call #54015 added in `kimi_k3/nvidia/mla.py`.

`vllm/model_executor/kernels/linear/cute_dsl/skinny_gemm.py` stays as an explicitly-invoked kernel. The two surviving fused-A dispatchers are folded onto a shared `vllm/models/common/ops/low_latency_linear.py` (~200 lines of duplication removed).

### 4. `--linear-backend` was being ignored by the LM head

`UnquantizedEmbeddingMethod.apply` and the compressed-tensors transform module called `dispatch_unquantized_gemm()` with no argument, so they always took `"auto"` regardless of the flag. Harmless before; a wrong kernel once `"auto"` means CuTeDSL — a GSM8K run with `--linear-backend=cutlass` still logged the CuTeDSL GEMM. All call sites now resolve the backend once per layer via `dispatch_configured_unquantized_gemm()`, which also lifts a per-forward dispatch off the LM head's hot path.

### 5. Sliced-weight GEMMs

MLA slices the fused QKV-A/gate weight so the halves can overlap on separate streams. The merged column-parallel projection stores one contiguous weight and the slices are taken along its rows, so both stay contiguous and 32-byte aligned and the CuTeDSL kernel claims them directly — verified for the Kimi-K3 TP8 and TP16 shapes. They just cannot go through the Linear module, not being layers, so they call the configured GEMM directly.

## Not a duplicate

- #50572 (merged) added the backend; this makes it the default and cleans up after it.
- #40638 is a broader draft integrating `mm_bf16`'s cuDNN/CUTLASS/tinygemm/cuBLASLt backends. Different scope: it adds backends, this one changes which backend `auto` picks for BF16 and removes the skinny-GEMM dispatch. They touch `dispatch_unquantized_gemm` in compatible ways.
- **Conflicts with #54447** ("Enable residual skinny GEMM on SM100"), which extends the residual skinny-GEMM path this PR removes. Flagging for the author — if that work is wanted, the residual path should come back as an explicit call rather than through the shape tables.

## Test plan and results

GB300 (SM103), `flashinfer-python==0.6.18`, branch at c0cab8ca30.

```
.venv/bin/python -m pytest tests/models/test_low_latency_linear.py \
    tests/kernels/test_bf16_skinny_gemm.py \
    tests/kernels/test_flashinfer_bf16_gemm.py -q
# 108 passed
```

`tests/kernels/test_bf16_skinny_gemm.py` is rewritten to drive the kernel directly (it has no model dispatch left to test): explicit configs, the heuristic config over M=1..16, the residual epilogue, CUDA graph capture, and the guard paths. `tests/models/test_low_latency_linear.py` is new and covers the fused-A tables, installation, capability routing, and the weight-slice chain.

### Model eval

GSM8K, 500 questions, 5-shot, greedy, `Qwen/Qwen3-0.6B` (dense BF16 — exercises the unquantized linear path end to end):

| linear-backend | CuTeDSL active | accuracy | invalid rate | output tokens |
|---|---|---|---|---|
| `cutlass` (baseline) | no | 0.428 | 0.002 | 50521 |
| `auto` (this PR) | yes | 0.428 | 0.002 | 50522 |

### Kernel benchmark: is CuTeDSL better than the fused-A GEMM everywhere?

No — which is why the fused-A tables stay. Added `benchmarks/kernels/benchmark_fused_a_vs_flashinfer_bf16.py`, which measures every `(N, K, M)` in the Kimi-K3 and GLM-5.2 fused-A tables on all three backends. CuTeDSL is measured after a FlashInfer autotuning pass (which profiles both its direct and split-K tactic spaces rather than using its shape heuristic). Timing is CUPTI kernel time under CUDA graph replay with a cold L2 — eager dispatch overhead dominates a CuTeDSL call by ~40x and would otherwise swamp the comparison.

Over 210 measured cases: **fused-A best in 175 (83%), CuTeDSL in 32 (15%), cuBLAS in 3**. Median CuTeDSL/fused-A speedup 0.91x (range 0.81x–1.19x). CuTeDSL wins on the large-N, K=7168 shapes and loses on everything with K ≤ 2048.

All 16 shapes, median over the M range (per-case JSON via `--json`):

| N | K | projection | M | cuBLAS µs | fused-A µs | CuTeDSL µs | CuTeDSL vs fused-A | winner |
|---:|---:|:---|:---:|---:|---:|---:|:---:|:---|
| 768 | 128 | f_b_proj (TP16) | 1-16 | 2.99 | 2.43 | 2.86 | 0.82-0.86x | fused-A |
| 1536 | 128 | f_b_proj | 1-16 | 3.04 | 2.50 | 2.94 | 0.81-0.88x | fused-A |
| 3072 | 128 | f_b_proj | 1-16 | 2.98 | 2.69 | 3.10 | 0.81-0.90x | fused-A |
| 7168 | 384 | shared_down_proj (TP16) | 1-8 | 4.32 | 4.18 | 4.42 | 0.89-1.00x | fused-A |
| 7168 | 768 | shared_down_proj | 1-16 | 5.57 | 5.34 | 5.78 | 0.89-0.97x | fused-A |
| 1152 | 1536 | q_b_proj (TP16) | 2-16 | 5.98 | 3.90 | 4.45 | 0.84-1.11x | mixed |
| 2304 | 1536 | q_b_proj | 1-16 | 5.82 | 4.48 | 5.25 | 0.84-1.09x | mixed |
| 4608 | 1536 | q_b_proj | 1-16 | 6.67 | 5.89 | 6.58 | 0.88-0.91x | fused-A |
| 2048 | 2048 | GLM-5.2 q_b_proj | 3-16 | 6.14 | 5.06 | 5.62 | 0.90-0.92x | fused-A |
| 2624 | 6144 | GLM-5.2 qkv_a_proj | 3-16 | 12.13 | 9.76 | 10.02 | 0.96-0.98x | fused-A |
| 768 | 7168 | mla_g_proj / shared_gate_up_proj (TP16) | 5-16 | 8.78 | 7.14 | 6.69 | 0.99-1.08x | mixed |
| 1536 | 7168 | shared_gate_up_proj / mla_g_proj | 1-16 | 10.93 | 7.97 | 8.18 | 0.95-1.00x | fused-A |
| 2112 | 7168 | fused_qkv_a_proj | 1-16 | 11.31 | 9.04 | 9.81 | 0.91-1.12x | mixed |
| 3216 | 7168 | in_proj_qkvgfab (TP16) | 9-15 | 16.00 | 15.20 | 12.83 | 1.18-1.19x | **CuTeDSL** |
| 3584 | 7168 | routed_expert_down_proj | 2-8 | 15.62 | 15.42 | 13.73 | 1.11-1.15x | **CuTeDSL** |
| 4224 | 7168 | dense_gate_up_proj (TP16) | 4-8 | 17.28 | 16.74 | 15.49 | 1.08-1.09x | **CuTeDSL** |

So the SM103 table is re-tuned against the new default: `3216x7168`, `3584x7168` and `4224x7168` are dropped entirely (CuTeDSL wins at every token count by 8-19%), and `768x7168`, `2112x7168`, `2304x1536`, `1152x1536` are trimmed where CuTeDSL took the low-M or high-M end. Every remaining `(shape, M)` is one the fused-A GEMM still wins, by 1.00x-1.24x (median 1.11x). The SM100 and SM90 tables are untouched — I have no B200 or H200 to re-measure on, and their entries are unaffected by an SM103 measurement.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan, such as providing test command.
- [x] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by vadiklyutiy** [comment]
@zyongye 
Below same measurements for B200

| N | K | projection | M | cuBLAS µs | fused-A µs | CuTeDSL µs | CuTeDSL vs fused-A | winner |
| --- | --- | --- | --- | --- | --- | --- | --- | --- |
| 768 | 128 | f_b_proj (TP16) | 1-16 | 3.17 | 2.51 | 3.04 | 0.79-0.84x | fused-A |
| 1536 | 128 | f_b_proj | 1-16 | 3.18 | 2.56 | 3.09 | 0.81-0.85x | fused-A |
| 3072 | 128 | f_b_proj | 1-16 | 3.20 | 2.78 | 3.26 | 0.82-0.87x | fused-A |
| 7168 | 384 | shared_down_proj (TP16) | 1-8 | 4.35 | 4.43 | 4.75 | 0.92-0.98x | mixed |
| 7168 | 768 | shared_down_proj | 1-16 | 5.63 | 5.63 | 5.89 | 0.90-0.99x | mixed |
| 1152 | 1536 | q_b_proj (TP16) | 2-16 | 5.15 | 4.13 | 4.64 | 0.80-1.18x | mixed |
| 2304 | 1536 | q_b_proj | 1-16 | 5.76 | 4.70 | 5.60 | 0.82-1.12x | mixed |
| 4608 | 1536 | q_b_proj | 1-16 | 6.37 | 6.08 | 6.78 | 0.87-0.92x | fused-A |
| 2048 | 2048 | GLM-5.2 q_b_proj | 3-16 | 6.42 | 5.22 | 5.97 | 0.87-0.93x | fused-A |
| 2624 | 6144 | GLM-5.2 qkv_a_proj | 3-16 | 12.40 | 10.11 | 10.26 | 0.95-0.99x | fused-A |
| 768 | 7168 | mla_g_proj / shared_gate_up_proj (TP16) | 5-16 | 9.26 | 7.49 | 6.85 | 1.03-1.10x | **CuTeDSL** |
| 1536 | 7168 | shared_gate_up_proj / mla_g_proj | 1-16 | 10.30 | 8.10 | 8.54 | 0.93-1.16x | mixed |
| 2112 | 7168 | fused_qkv_a_proj | 1-16 | 11.74 | 9.42 | 10.05 | 0.92-1.14x | mixed |
| 3216 | 7168 | in_proj_qkvgfab (TP16) | 9-15 | 16.48 | 15.90 | 13.34 | 1.17-1.21x | **CuTeDSL** |
| 3584 | 7168 | routed_expert_down_proj | 2-8 | 16.13 | 15.94 | 14.21 | 1.11-1.38x | **CuTeDSL** |
| 4224 | 7168 | dense_gate_up_proj (TP16) | 4-8 | 17.79 | 17.25 | 16.19 | 1.05-1.07x | **CuTeDSL** |

**Comment by vadiklyutiy** [comment]
@zyongye pls look on https://github.com/flashinfer-ai/flashinfer/pull/4908
seems after it we can fully avoid using fused-A gemm because FI cutedsl is always better

**Comment by NolenLiang** [comment]
GB200 (SM100) data for Qwen3.8-Flash-Next-FP8 that may be useful for this PR. **Scope:** these runs measure the `--linear-backend flashinfer_cutedsl` flag on the nightlies named below, not this PR's code (the PR conflicts with main and its files have moved since, so I could not overlay it on a nightly image).

Setup: one tray of 4× GB200, `--tensor-parallel-size 4 --enable-expert-parallel`, vLLM nightly `2a02f6e` (2026-09-10) unless noted, vllm-bench random 8192-in / 1024-out, unique prompts per point, 8·c requests per point, streaming. This checkpoint keeps every dense projection (GDN in/out proj, QSA q/k/v/o + indexer, hyper-connection, shared experts, gates) in BF16 (`modules_to_not_convert`), so at decode the ~15 M ≤ 16 BF16 GEMMs per layer go through cuBLAS split-K plus a separate reduce kernel (~484 + 242 launches per step at c=1) because `qwen4_exp/nvidia/low_latency_gemm.py` has no SM100 table.

| conc | stock (cuBLAS split-K) tok/s | `--linear-backend flashinfer_cutedsl` tok/s | change |
|---|---|---|---|
| 1 | 183 | 193 | +5.6 % |
| 2 | 307 | 335 | +9.0 % |
| 4 | 572 | 609 | +6.4 % |
| 8 | 998 | 1056 | +5.8 % |
| 16 | 1669 | 1713 | +2.6 % |

Two further same-tray A/B rounds over c=1–128 (fresh server per run) reproduced this: +5.3…5.7 / +9.2…9.3 / +6.3…8.7 / +3.4…5.0 / +3.1…4.9 / +2.0…3.6 % at c=1/2/4/8/16/32 and ±0.3 % at c=64/128 (run-to-run spread ≤ 2 %). Two caveats from the same runs: (1) mean TTFT rises with the flag at low concurrency (+15…60 ms on a 172–384 ms base at c ≤ 4, unchanged at c ≥ 64), so it is a decode-throughput win rather than a no-regression change; (2) the CuTeDSL GEMM ran with its default configuration (`--no-enable-flashinfer-autotune`, as in the vLLM recipe); enabling autotune added another +6.4 / +6.2 % at c=1–2 and ≈ 0 at c ≥ 4 in a single round, at the cost of server start-up going from ~200 s to ~700 s. gsm8k (5-shot greedy, chat mode with the qwen3 reasoning parser) with and without the flag: 0.915 vs 0.917 flexible-extract, 26 vs 29 discordant questions of 1319, paired exact McNemar p = 0.79.

Co-developed with an AI assistant; measurements run and checked by the author.
